src/Temporalio/Bridge/src/worker.rs

```diff
@@ -56,27 +63,231 @@ pub struct TunerHolder {
 }
 
 #[repr(C)]
-#[derive(Clone, Copy, PartialEq)]
+#[derive(Clone, Copy)]
 pub enum SlotSupplier {
     FixedSize(FixedSizeSlotSupplier),
     ResourceBased(ResourceBasedSlotSupplier),
+    Custom(CustomSlotSupplierCallbacksImpl),
 }
 
 #[repr(C)]
-#[derive(Clone, Copy, PartialEq)]
+#[derive(Clone, Copy)]
 pub struct FixedSizeSlotSupplier {
     num_slots: usize,
 }
 
 #[repr(C)]
-#[derive(Clone, Copy, PartialEq)]
+#[derive(Clone, Copy)]
 pub struct ResourceBasedSlotSupplier {
     minimum_slots: usize,
     maximum_slots: usize,
     ramp_throttle_ms: u64,
     tuner_options: ResourceBasedTunerOptions,
 }
 
+#[repr(C)]
+pub struct CustomSlotSupplier<SK> {
+    inner: CustomSlotSupplierCallbacksImpl,
+    _pd: std::marker::PhantomData<SK>,
+}
+
+unsafe impl<SK> Send for CustomSlotSupplier<SK> {}
+unsafe impl<SK> Sync for CustomSlotSupplier<SK> {}
+
+type CustomReserveSlotCallback =
+    unsafe extern "C" fn(ctx: SlotReserveCtx, sender: *mut libc::c_void);
+type CustomCancelReserveCallback = unsafe extern "C" fn(token_source: *mut libc::c_void);
+/// Must return C#-tracked id for the permit. A zero value means no permit was reserved.
+type CustomTryReserveSlotCallback = unsafe extern "C" fn(ctx: SlotReserveCtx) -> usize;
+type CustomMarkSlotUsedCallback = unsafe extern "C" fn(ctx: SlotMarkUsedCtx);
+type CustomReleaseSlotCallback = unsafe extern "C" fn(ctx: SlotReleaseCtx);
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct CustomSlotSupplierCallbacksImpl(*const CustomSlotSupplierCallbacks);
```

Comment:
IMO this needs a `Drop` impl that calls a separate callback called `free` that invokes `Free` on the .NET side similar to how the custom metric meter does.

Reply:
Yes, just that was the last thing on my TODO list. Added.